### DISCUSSION TITLE
Multiarch Docker

### DISCUSSION THREAD

**Discussion by yarons** [nextcloud/all-in-one#490] [Ideas] [answered]
Docker supports multiarchitecture build nowadays, which means you can upload several docker architecture variants for the same tag to Docker hub.

This can all be done using a well configured Docker's buildx.

I have a working example on my GH repo which uploads the multiarch artifact to ghcr.io instead of Docker hub:
https://github.com/yarons/pritunl-zero-docker/actions/runs/1736367592/workflow
https://github.com/yarons/pritunl-zero-docker/blob/main/.github/workflows/docker-image.yml

These are the actual artifacts (ready for pull in both architectures):
https://github.com/yarons?tab=packages&repo_name=pritunl-zero-docker

**Comment by szaimen** [comment]
Hi, thanks for your idea! Unfortunately we cannot build multiarch images as we rely on the docker repository api which itself is not able to handle multiarch container images. See https://github.com/nextcloud/all-in-one/pull/42#issuecomment-986057315
So this is unfortunately a won't fix for the time being.

**Comment by yarons** [comment]
@szaimen So how do you explain this? (Or am I missing something?)
https://hub.docker.com/_/mongo?tab=tags

**Comment by szaimen** [comment]
Yes, docker hub is able to display multiarch images. I don't know if they implemented a custom, not documented api or how they are doing it. Feel free to investigate how you can retreive the checksums for multiarch images. See https://github.com/nextcloud/all-in-one/pull/42#issuecomment-986051622 for some initial steps

**Comment by szaimen** [comment]
If you find a way, we may consider building multiarch images.

**Comment by yarons** [comment]
@szaimen The pipelines in my GitHub are modified for ghcr but I can adapt these to Dockerhub (The original script was for Dockerhub but I prefer ghcr so I modified it), where are your pipelines running?

If we were to modify my pipelines all you need to do is to replace the name of the image to dockerhub instead of ghcr and login to Dockerhub (The username and password should be stored as Docker secrets), then run docker push.

**Comment by szaimen** [comment]
No, that is not what I am talking about. Did you read the comments that I've linked to?

**Comment by szaimen** [comment]
The problem is that we are not able to retreive the checksums of multiarch images via the docker repo api

**Comment by szaimen** [comment]
I know that it is possible to create multiarch container images without much trouble

**Comment by yarons** [comment]
I see but why do you care so much for the hashes? Docker is agile enough to pull the right image to the right hosting arch.

**Comment by szaimen** [comment]
Because we need that in order to be able to judge if the containers are up-to-date or not ;)

**Comment by szaimen** [comment]
See e.g. https://github.com/nextcloud/all-in-one/blob/4e852761d91a3bc764ef389906c27e9504a66259/php/src/Docker/DockerActionManager.php#L73-L85

**Comment by yarons** [comment]
I saw something about tag architecture in here, could that possibly help?
https://www.arthurkoziel.com/dockerhub-registry-api/
(Get Image Manifest section).

**Comment by szaimen** [comment]
Thanks but that is kind of actually what we are doing already: see https://github.com/nextcloud/all-in-one/blob/4e852761d91a3bc764ef389906c27e9504a66259/php/src/Docker/DockerHubManager.php#L37-L45

So as I said: until you show me a way to get all needed hashes of multiarch images, this will not be implemented :)

**Comment by yarons** [comment]
I see, it looks like although there's architecture field in his responsed I couldn't find that officially.

I'll get back to you if I'll find something, thanks.

**Comment by szaimen** [comment]
Thank you, too! :)

**Comment by yarons** [comment]
I think I found something, run the following command on your machine and look at the output:
docker manifest inspect --verbose library/ubuntu:latest

There must be a way, I'll try to see if I can get some documentation for it but I can definitely see multiarch hashes.

On the docker docs there's a warning saying that this is experimental.

**Comment by szaimen** [comment]
Hi, I just found out that it is possible to get all digest by querying this endpoint:

https://registry.hub.docker.com/v2/repositories/user/repository/tags

<details>
<summary>example</summary>

```
curl -L -s 'https://registry.hub.docker.com/v2/repositories/library/centos/tags?page_size=1024'

{
  "creator": 1156886,
  "id": 105191642,
  "images": [
    {
      "architecture": "amd64",
      "features": "",
      "variant": null,
      "digest": "sha256:fd84102fc72960dd1b8da0ee3b4c13e3b0c1d2a085de118bc4c97821cd986e02",
      "os": "linux",
      "os_features": "",
      "os_version": null,
      "size": 74866818,
      "status": "active",
      "last_pulled": "2023-01-05T20:52:45.259516Z",
      "last_pushed": null
    },
    {
      "architecture": "ppc64le",
      "features": "",
      "variant": null,
      "digest": "sha256:77eebe55df5e0521188b7fa224feddb1877b8c40901173e0f87aa3a97ecef4d2",
      "os": "linux",
      "os_features": "",
      "os_version": null,
      "size": 84620966,
      "status": "active",
      "last_pulled": "2023-01-05T21:09:32.02647Z",
      "last_pushed": null
    },

```

</details>

However creating multi-arch images will also mean that testing gets much slower since I would alway need to wait until the multi-arch image is built which takes up to 45min compared to around 5min since Github Actions does not allow for hardware-accellaration. So I'd rather wait for that to be available before taking another look at this.

**Reply by yarons** [reply]
It's possible to do it in parallel, I think that this is the default behavior in `buildx`.
Sorry, it also applies to the testing which has nothing to do with `buildx`.

**Reply by szaimen** [reply]
So does each arch become available directly on docker hub when it is built? For me it looks like they are first built and afterwards pushed to docker hub together as multiarch image...

**Reply by Zoey2936** [reply]
> So does each arch become available directly on docker hub when it is built? For me it looks like they are first built and afterwards pushed to docker hub together as multiarch image...

amd64 and arm64 will first get pushed when both builds are ready

**Reply by szaimen** [reply]
> amd64 and arm64 will first get pushed when both builds are ready

Yeah, I thought so. Lets wait for hardware-acvelleration becoming available for github workflows then.

**Reply by yarons** [reply]
Basically the "multiarch" approach is merely adding a single tag to 2 different hashes, if there's a way to modify tag after its being created then I can possibly add the first hash first and the second hash when it's ready, this way it'll take longer but at least one of the builds will be available.

**Reply by Zoey2936** [reply]
> Basically the "multiarch" approach is merely adding a single tag to 2 different hashes, if there's a way to modify tag after its being created then I can possibly add the first hash first and the second hash when it's ready, this way it'll take longer but at least one of the builds will be available.

I am not sure, but it could be possible to do this using regctl(https://github.com/regclient/regclient/blob/main/docs/regctl.md)

**Reply by Zoey2936** [reply]
Or using this part: „The hard way with docker manifest“ from here: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

**Reply by Zoey2936** [reply]
`docker manifest create nextcloud/all-in-one:latest -a nextcloud/all-in-one:latest -a nextcloud/all-in-one:latest-arm64`
`docker manifest push nextcloud/all-in-one:latest`
could work

**Reply by szaimen** [reply]
What about keeping the build like it is and only pushing the multiarch to beta and latest by using docker manifest create? This could probably be added here https://github.com/nextcloud-releases/all-in-one/blob/9b649b6e1705abafb95b1828bb873632ec318e7d/.github/workflows/promote-to-beta.yml#L50 and and  here https://github.com/nextcloud-releases/all-in-one/blob/9b649b6e1705abafb95b1828bb873632ec318e7d/.github/workflows/promote-to-latest.yml#L83

**Comment by szaimen** [comment]
TODO:
- [x] merge and test https://github.com/nextcloud-releases/all-in-one/pull/31
- [x] https://github.com/nextcloud/all-in-one/pull/1887
- [x] adjust the documentation (only when latest multiarch is provided)

**Reply by Zoey2936** [reply]
@szaimen do you maybe know the differents between these apis? Or are they the same?
https://hub.docker.com/v2/repositories/library/centos/tags?page_size=1024
https://registry.hub.docker.com/v2/repositories/library/centos/tags?page_size=1024
https://hub.docker.com/v2/namespaces/library/repositories/centos/tags?page_size=1024

**Reply by szaimen** [reply]
I am honestly not sure. Probably we can use any of them...

**Reply by Zoey2936** [reply]
Even if I think this should work, I don't understand how watchtower works with digests of multiarch, since they use another API (so it works with repos outside docker hub)

**Reply by szaimen** [reply]
yeah, not sure either

**Reply by szaimen** [reply]
I think they have multiple techniques. In worst case they fall back to manual pulling a new image iirc...

**Reply by Zoey2936** [reply]
https://docs.docker.com/registry/spec/manifest-v2-2/ was the problem maybe that the mastercontainer didnt accepted the "application/vnd.docker.distribution.manifest.list.v2+json" manifest?

**Reply by szaimen** [reply]
I think it is fine for now to only accept docker hub images. We can later on still improve it.

**Comment by szaimen** [comment]
So basically everything was merged and built correclty but the digest did not match:
https://github.com/nextcloud-releases/all-in-one/pull/31#issuecomment-1413963952

So not sure how to proceed from here. I reverted the multiarch building and now the update check works correctly again. So it might depend on how exactly the manifests get created but not sure how to retrieve the correct digest. This needs more testing.

**Comment by szaimen** [comment, marked as answer]
This is now released with v4.3.1 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
